app/abci.go

```diff
@@ -58,8 +56,9 @@ func (app *App) PreprocessTxs(txs abci.RequestPreprocessTxs) abci.ResponsePrepro
 			continue
 		}
 
+		squareSize := wireMsg.MessageShareCommitment[0].K
 		// parse wire message and create a single message
-		coreMsg, unsignedPFM, sig, err := types.ProcessWirePayForMessage(wireMsg, app.SquareSize())
+		coreMsg, unsignedPFM, sig, err := types.ProcessWirePayForMessage(wireMsg, squareSize)
```

Comment:
while this seems logical at first, we actually need all messages for a block to be processed using the same correct square size. If this doesn't happen, then messages and their corresponding `MsgPayForMessage`s in the square will be considered invalid.

We actually don't have a mechanism to pick the square size (see [#454](https://github.com/celestiaorg/celestia-core/issues/454)), so we delegated that functionality to a method that we plan on updating later.

Reply:
also see https://github.com/celestiaorg/celestia-specs/blob/master/src/rationale/message_block_layout.md

Reply:
Thank you so much for the info, Evan! This helps me to clarify a lot.
I have some follow-up questions.
- In the proposal `PreProcessTxsRequest`, the final square size that be used in celestia-core is `square_width`?
- "The user, therefore, submits multiple signatures for different potential square sizes in order to ensure that one will be valid for whatever the square size ends up being for the upcoming block": the square-sizes input from flag will be a list of potential square sizes? What happens if the user submitted multiple signatures (different square sizes) but there is no signature for the correct square size of the upcoming block?

Reply:
The square size is not currently communicated between the app and celestia-core (fork of tendermint) see [#454](https://github.com/celestiaorg/celestia-core/issues/454) (this is changing soon) this is why we simply use the max square size for each block.

> the square-sizes input from flag will be a list of potential square sizes?

yeah, and it should be passed to the [function](https://github.com/celestiaorg/celestia-app/blob/63519eccc053e9695af5901164dfd5da5fe3fb24/x/payment/types/wirepayformessage.go#L18-L40) that creates wire pay for messages, and that's pretty much all that's needed for this PR. That and adding multiple square sizes to the cli integration test

the `ValidateBasic` method for `MsgWirePayForMessage` should check that the signatures are present. This method is called (twice iirc) before being added to the mempool. If there is no sig, then the tx is not valid and will not be propogated or make it into the proposal block of an honest proposer.

Reply:
Thanks, Evan, I have a better understanding of what happens.